File: exo4.2.py
import os
import sys

# Programme qui lie stdout et stdin du père à un tube

def main():
    r, w = os.pipe()
    pid = os.fork()

    if pid > 0:  # Processus père
        os.close(r)
        # Sauvegarder le stdout original
        original_stdout = os.dup(1)
        # Rediriger stdout vers le tube
        os.dup2(w, 1)
        
        print("Message_via_printf :")
        
        # Restaurer stdout
        os.dup2(original_stdout, 1)
        os.close(w)
        os.close(original_stdout)
    else:  # Processus fils
        os.close(w)
        # Rediriger stdin depuis le tube
        original_stdin = os.dup(0)
        os.dup2(r, 0)
        
        buffer = input()
        print(f"Le fils a recu : {buffer}")
        
        # Restaurer stdin
        os.dup2(original_stdin, 0)
        os.close(r)
        os.close(original_stdin)

if __name__ == "__main__":
    main()

File: exo4.2.c
#include <stdio.h>
#include <unistd.h>

/*(R) eprendre les programmes de l’exercice ??. Nous allons faire un programme
qui fait la même chose, mais transmet les données diﬀérement. Dans le programme père, on
liera les ﬂots stdout et stdin à un tube.*/
int main () {

    int tube[2];

    pipe (tube);

    if (fork () > 0) {
        close (tube[0]);
        dup2 (tube[1], STDOUT_FILENO);

        printf ("Message_via_printf : ");
        close (tube[1]);
    } else {
        close (tube[1]);

        dup2 (tube[0], STDIN_FILENO);

        char buffer[256];
        scanf ("%s", buffer);
        printf ("Le fils a recu : %s\n", buffer);
        close (tube[0]);
    }

    return 0;
}